File: swift/300_LongestIncreasingSubsequence.swift
class Solution {
    func lengthOfLIS(_ nums: [Int]) -> Int {
        var monotonicIncreasingList = [Int]()
        for n in nums {
            if monotonicIncreasingList.isEmpty || n > monotonicIncreasingList.last! {
                monotonicIncreasingList.append(n)
            } else {
                for i in 0..<monotonicIncreasingList.count {
                    if n < monotonicIncreasingList[i] && (i == 0 || n > monotonicIncreasingList[i - 1]) {
                        monotonicIncreasingList[i] = n
                        break
                    }
                }
            }
        }
        return monotonicIncreasingList.count
    }
}

File: python/300_LongestIncreasingSubsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        mono = []
        for n in nums:
            if not mono or n > mono[-1]:
                mono.append(n)
            else:
                for i, m in enumerate(mono):
                    if n < m and (i == 0 or n > mono[i - 1]):
                        mono[i] = n
                        break
        return len(mono)
